Source/Provers/SMTLib/ProverInterface.cs

```diff
@@ -347,15 +388,80 @@ private void PrepareCommon()
+        Dictionary<Function, VCExpr> functionDefinitionMap = new Dictionary<Function, VCExpr>();
+        Stack<Function> functionDefs = new Stack<Function>();
+        Stack<VCExpr> otherAxioms = new Stack<VCExpr>();
+        for (int i = axiomList.Count-1; i >= 0; i--) {
+          Function f = IsFunctionDef(axiomList[i]);
+          if (f != null) {
+            // Add as a known function to DeclCollector so that it's not declared later.
+            DeclCollector.AddFunction(f);
+            functionDefinitionMap.Add(f, axiomList[i]);
+            functionDefs.Push(f);
+          } else {
+            otherAxioms.Push(axiomList[i]);
+          }
+        }
+
+        // Process each definition, but also be sure to process dependencies first in case one definition calls another.
+        // Also check for definition cycles.
+        FunctionDependencyCollector collector = new FunctionDependencyCollector();
+        HashSet<Function> axiomAdded = new HashSet<Function>(); // whether definition has been fully processed
+        HashSet<Function> dependenciesComputed = new HashSet<Function>(); // whether dependencies have already been computed
+        while (functionDefs.Count > 0) {
+          Function f = functionDefs.Peek();
+          if (axiomAdded.Contains(f)) {
+            // This definition was already processed (as a dependency of another definition)
+            functionDefs.Pop();
+            continue;
+          }
+          // Grab the definition and then compute the dependencies.
+          Contract.Assert(functionDefinitionMap.ContainsKey(f));
+          VCExpr expr = functionDefinitionMap[f];
+          var quan = expr as VCExprQuantifier;
+          Contract.Assert(quan != null && quan.Infos.isFunctionDefinition);
+          var body = quan.Body as VCExprNAry;
+          List<Function> dependencies = collector.Collect(body[1]);
+          bool hasDependencies = false;
+          foreach (Function fdep in dependencies) {
+            if (functionDefinitionMap.ContainsKey(fdep) && !axiomAdded.Contains(fdep)) {
+              if (!dependenciesComputed.Contains(fdep)) {
+                // Handle dependencies first
+                functionDefs.Push(fdep);
+                hasDependencies = true;
+              } else {
+                HandleProverError("Function definition cycle detected: " + f.ToString() + " depends on " + fdep.ToString());
+              }
+            }
+          }
+          if (!hasDependencies) {
+            // No dependencies: go ahead and process this definition.
+            AddAxiom(VCExpr2String(expr, -1));
+            axiomAdded.Add(f);
+            functionDefs.Pop();
+          } else {
+            dependenciesComputed.Add(f);
+          }
+        }
+
+	// Now process the rest of the axioms.
+        while (otherAxioms.Count > 0) {
+          VCExpr expr = otherAxioms.Pop();
+          var str = VCExpr2String(expr, -1);
+          if (str != "true")
+            AddAxiom(str);
```

Comment:
Coding style: always use curly braces around the `if` branches

Reply:
Fixed here and in a few other places.